In [1]:
import random
import math
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score, precision_score, recall_score

#### Importando o dataset e criando os dados de treino e teste

In [3]:
df_breast_cancer = pd.read_csv("breast-cancer.csv")
df_breast_cancer

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0    ...        25.380          17.33           184.60      2019.0   
1    ...        24.990          23.41           158.80      1956.0   
2    ...        23.570          25.53           152.50      1709.0   
3    ...        14.910          26.50            98.87       567.7   
4    ...        22.540          16.67           152.20      1575.0   
..   ...           ...            ...              ...         ...   
564  ...        25.450          26.40           166.10      2027.0   
565  ...        23.690          38.25           155.00      1731.0   
566  ...        18.980          34.12           126.70      1124.0   
567  ...        25.740          39.42           184.60      1821.0   
568  ...         9.456          30.37            59.16       268.6   

     smoothness_worst  compactness_worst  concavity_worst  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
0                  0.2654          0.4601                  0.11890  
1                  0.1860          0.2750                  0.08902  
2                  0.2430          0.3613                  0.08758  
3                  0.2575          0.6638                  0.17300  
4                  0.1625          0.2364                  0.07678  
..                    ...             ...                      ...  
564                0.2216       

In [4]:
radius_mean = df_breast_cancer["radius_mean"].values
area_mean	 = df_breast_cancer["area_mean"].values
perimeter_mean = df_breast_cancer["perimeter_mean"].values
concave_points_mean = df_breast_cancer["concave points_mean"].values

diagnosis = df_breast_cancer["diagnosis"].values

Vamos colocar os atributos em uma matriz e os target em uma lista

In [5]:
atributos = [[i, j, k, l] for i, j, k, l in zip(radius_mean, area_mean, perimeter_mean, concave_points_mean)]
target = diagnosis

#### vamos agora aplicar o Label Encoder para dados binários:

##### 1 - para tumores malignos
##### 0 - para tumores benigno

In [8]:
target = target.reshape(-1,1)


In [9]:
encoder = LabelEncoder()
codificados = encoder.fit_transform(target)

C:\Users\glauber24022\AppData\Roaming\Python\Python311\site-packages\sklearn\preprocessing\_label.py:114: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Como o dataset é imenso, vou fazer dois splits de dados para ter apenas uma porcentagem do dataset

In [6]:
len(target)

569

In [11]:
X_treino, X_teste, y_treino, y_teste = train_test_split(atributos, codificados, test_size=0.1, random_state=2020)

len(y_treino)

512

In [12]:
X_treino, X_teste, y_treino, y_teste = train_test_split(X_teste, y_teste, test_size=0.2, random_state=2020)

len(y_treino)

45

nomalizando o dados de treino

In [13]:
scaler = StandardScaler() #normalizador padrão

X_treino_norm = scaler.fit_transform(X_treino)

In [14]:
X_teste_norm = scaler.transform(X_teste)

#### Classe Valor

A classe valor foi baseada de acordo com a classe realizada no vídeo do professor e no notebook ATP-303 NN 4.2 - Notebook MLP.


Foi necessário a implementação de novos métodos para o funcionamento correto da classe Valor no resto do código.

In [15]:
class Valor:
    def __init__(self, data, progenitor=(), operador_mae="", rotulo=""):
        self.data = data
        self.progenitor = progenitor
        self.operador_mae = operador_mae
        self.rotulo = rotulo
        self.grad = 0

    def __repr__(self):
        return f"Valor(data={self.data})"
    
    def __add__(self, outro_valor):
        """Realiza a operação: self + outro_valor."""
        
        if not isinstance(outro_valor, Valor):
            outro_valor = Valor(outro_valor)
            
        progenitor = (self, outro_valor)
        data = self.data + outro_valor.data
        operador_mae = "+"
        resultado = Valor(data, progenitor, operador_mae)
        
        def propagar_adicao():
            self.grad += resultado.grad
            outro_valor.grad += resultado.grad
            
        resultado.propagar = propagar_adicao
        
        return resultado
    
    def __mul__(self, outro_valor):
        """Realiza a operação: self * outro_valor."""
        
        if not isinstance(outro_valor, Valor):
            outro_valor = Valor(outro_valor)
            
        progenitor = (self, outro_valor)
        data = self.data * outro_valor.data
        operador_mae = "*"
        resultado = Valor(data, progenitor, operador_mae)
        
        def propagar_multiplicacao():
            self.grad += resultado.grad * outro_valor.data 
            outro_valor.grad += resultado.grad * self.data
            
        resultado.propagar = propagar_multiplicacao
        
        return resultado
    
    def exp(self):
        """Realiza a operação: exp(self)"""
        progenitor = (self, )
        data = math.exp(self.data)
        operador_mae = "exp"
        resultado = Valor(data, progenitor, operador_mae)
        
        def propagar_exp():
            self.grad += resultado.grad * data 
        
        resultado.propagar = propagar_exp
        
        return resultado
    
    def __pow__(self, expoente):
        """Realiza a operação: self ** expoente"""
        assert isinstance(expoente, (int, float))
        progenitor = (self, )
        data = self.data ** expoente
        operador_mae = f"**{expoente}"
        resultado = Valor(data, progenitor, operador_mae)
        
        def propagar_pow():
            self.grad += resultado.grad * (expoente * self.data ** (expoente - 1))
        
        resultado.propagar = propagar_pow
        
        return resultado
    
    def __truediv__(self, outro_valor):
        """Realiza a operação: self / outro_valor"""
        return self * outro_valor ** (-1)
    
    def __neg__(self):
        """Realiza a operação: -self"""
        return self * -1
    
    def __sub__(self, outro_valor):
        """Realiza a operação: self - outro_valor"""
        return self + (-outro_valor)
    
    def __rsub__(self, outro):
        if not isinstance(outro, Valor):
            outro = Valor(outro)
        return Valor(outro.data - self.data)
    
    def __radd__(self, outro_valor):
        """Realiza a operação: outro_valor + self"""
        return self + outro_valor
    
    def __rmul__(self, outro_valor):
        """Realiza a operação: outro_valor * self"""
        return self * outro_valor
    
    def sig(self):
        """Realiza a operação: exp(self) / (exp(self) + 1)"""
        return self.exp() / (self.exp() + 1)
    
    def log(self):
        """Realiza a operação: Log(self.data)"""
        return Valor(math.log(self.data))
    
    def propagar(self):
        pass

    def __gt__(self, outro_valor):
        """Realiza a operação maior que"""
        if not isinstance(outro_valor, Valor):
            outro_valor = Valor(outro_valor)
        return self.data > outro_valor.data

    def __lt__(self, outro_valor):
        """Realiza a operação menor que"""
        if not isinstance(outro_valor, Valor):
            outro_valor = Valor(outro_valor)
        return self.data < outro_valor.data
    
    def __int__(self):
        """Realiza a transformação para inteiro"""
        return int(self.data)
    
    def propagar_tudo(self):
        
        self.grad = 1
        
        ordem_topologica = []
        
        visitados = set()

        def constroi_ordem_topologica(v):
            if v not in visitados:
                visitados.add(v)
                for progenitor in v.progenitor:
                    constroi_ordem_topologica(progenitor)
                ordem_topologica.append(v)

        constroi_ordem_topologica(self)
        
        for vertice in reversed(ordem_topologica):
            vertice.propagar()

#### Classe Neurônio

A classe Neurônio foi baseada de acordo com a classe realizada no vídeo do professor e no notebook ATP-303 NN 4.2 - Notebook MLP.


Não foi necessário a implementação de novos métodos para o funcionamento correto da classe Neurônio no resto do código.

In [17]:
class Neuronio:
    def __init__(self, num_dados_entrada):
        self.vies = Valor(random.uniform(-1, 1))
        
        self.pesos = []
        for i in range(num_dados_entrada):
            self.pesos.append(Valor(random.uniform(-1, 1)))
            
    def __call__(self, x):
        
        assert len(x) == len(self.pesos)
        
        soma = 0
        for info_entrada, peso_interno in zip(x, self.pesos):
            soma += info_entrada * peso_interno
            
        soma += self.vies  
        dado_de_saida = soma.sig()
        
        return dado_de_saida       
    
    def parametros(self):
        return self.pesos + [self.vies]

#### Classe Camada

A classe Camada foi baseada de acordo com a classe realizada no vídeo do professor e no notebook ATP-303 NN 4.2 - Notebook MLP.


Não foi necessário a implementação de novos métodos para o funcionamento correto da classe Valor no resto do código.

In [16]:
class Camada:
    def __init__(self, num_neuronios, num_dados_entrada):
        neuronios = []
        
        for _ in range(num_neuronios):
            neuronio = Neuronio(num_dados_entrada)
            neuronios.append(neuronio)
            
        self.neuronios = neuronios     
        
    def __call__(self, x):
        dados_de_saida = []
        
        for neuronio in self.neuronios:
            informacao = neuronio(x)
            dados_de_saida.append(informacao)
            
        if len(dados_de_saida) == 1:
            return dados_de_saida[0]
        else:        
            return dados_de_saida  
    
    def parametros(self):
        params = []
        
        for neuronio in self.neuronios:
            params_neuronio = neuronio.parametros()
            params.extend(params_neuronio)
        
        return params

#### MLP


A classe MLP foi baseada de acordo com a classe realizada no vídeo do professor e no notebook ATP-303 NN 4.2 - Notebook MLP.


Não foi necessário a implementação de novos métodos para o funcionamento correto da classe no resto do código.

In [18]:
class MLP:
    def __init__(self, num_dados_entrada, num_neuronios_por_camada):
        
        percurso = [num_dados_entrada] + num_neuronios_por_camada
        
        camadas = []
        
        for i in range(len(num_neuronios_por_camada)):
            camada = Camada(num_neuronios_por_camada[i], percurso[i])
            camadas.append(camada)
            
        self.camadas = camadas
        
    def __call__(self, x):
        for camada in self.camadas:
            x = camada(x)
        return x
    
    def parametros(self):
        params = []
        
        for camada in self.camadas:
            parametros_camada = camada.parametros()
            params.extend(parametros_camada)
            
        return params

#### Treinando o modelo

Os dados apresentados abaixo fazem parte da arquitetura da Rede Neural

In [20]:
NUM_DADOS_DE_ENTRADA = 4  
NUM_DADOS_DE_SAIDA = 1    
CAMADAS_OCULTAS = [4, 2]  

arquitetura_da_rede = CAMADAS_OCULTAS + [NUM_DADOS_DE_SAIDA]

mlp_diagnostica = MLP(NUM_DADOS_DE_ENTRADA, arquitetura_da_rede)

Para Redes Neurais Classificadoras, o que muda da Rede Neural Regressora é justamente a função de perda.

Nesse caso, vou usar a função Binary Cross Entropy (BCE)(Entropia Cruzada) para a função de perda.

O papel da função de custo é retornar valores altos para previsões ruins e valores baixos para previsões boas. Por isso, queremos chegar o mais próximo do 0 possivel!

Mas por que vamos usar a BCE?

Pois ele lida bem com probabilidades, penaliza as previsões erradas com muita confiança e é diferenciável, ou seja, permite o cálculo do gradiente.

In [137]:
def binary_cross_entropy(y_true, y_pred):
    # Adiciona epsilon para evitar log(0)
    epsilon = 1e-7

    # Clamping manual (Valor não aceita min/max direto)
    y_pred = y_pred if y_pred > epsilon else Valor(epsilon)
    y_pred = y_pred if y_pred < (1 - epsilon) else Valor(1 - epsilon)

    # Cálculo da perda usando operadores da classe Valor
    loss = - (y_true * y_pred.log() + (1 - y_true) * (1 - y_pred).log())
    return loss


In [138]:
erros = []

for yt, yp in zip(y_treino, y_pred):
    BCE = binary_cross_entropy(yt, yp)
    erros.append(BCE)
    
loss = sum(erros)
print(loss)

Valor(data=31.167588618636717)


In [141]:
NUM_EPOCAS = 200
TAXA_DE_APRENDIZADO = 0.5

for epoca in range(NUM_EPOCAS):
    # forward pass
    y_pred = []
    for exemplo in X_treino_norm:
        previsao = mlp_diagnostica(exemplo)
        y_pred.append(previsao)

    # loss
    erros = []
    for yt, yp in zip(y_treino, y_pred):
        residuo = yp - yt
        erro_quadratico = residuo ** 2
        erros.append(erro_quadratico)        
    loss = sum(erros)

    # zero grad
    for p in mlp_diagnostica.parametros():
        p.grad = 0

    # backpropagation
    loss.propagar_tudo()

    # atualiza parâmetros
    for p in mlp_diagnostica.parametros():
        p.data = p.data - p.grad * TAXA_DE_APRENDIZADO

    # mostra resultado (opcional)
    print(epoca, loss.data)

0 11.238006310143245
1 11.434178831203335
2 14.875198013488703
3 17.786083686795312
4 11.489202402892158
5 14.181682171896675
6 15.559141918404364
7 12.610849682307617
8 15.20672216361952
9 13.062553628380893
10 14.070494480255737
11 14.171171472874539
12 3.8631290979964183
13 5.87695352388899
14 2.5844523525170833
15 3.0942722719621027
16 3.4290117632323533
17 4.806777761612066
18 2.583442854401301
19 2.8349090940645247
20 3.443004333517494
21 2.0859873645765825
22 2.337224006179788
23 2.5790043519658474
24 2.7632406989335405
25 2.475722434534464
26 2.5146647468601326
27 2.5190925662285717
28 2.3521536957069156
29 2.5011536805228034
30 2.17212038191921
31 2.4356676614721886
32 2.04493698970238
33 2.3498239499646725
34 1.9973055190611928
35 2.3096316264563157
36 1.974782856824933
37 2.288560704488177
38 1.9574571906782974
39 2.2692280363447153
40 1.9475178483570177
41 2.2540988812931184
42 1.9401760132697796
43 2.2384188245473258
44 1.9373987136851096
45 2.2236675058572892
46 1.9370516

In [144]:
y_pred_teste = []
for exemplo in X_teste_norm:
    previsao = mlp_diagnostica(exemplo)
    y_pred_teste.append(previsao)

In [145]:
print(y_pred_teste)

[Valor(data=0.9904898571586706), Valor(data=0.9905239336350088), Valor(data=0.028603679892504293), Valor(data=0.9905223572416822), Valor(data=0.02894576081738718), Valor(data=0.028472654681825173), Valor(data=0.9904651523252445), Valor(data=0.038554219997341176), Valor(data=0.9905270912218895), Valor(data=0.02848150283265805), Valor(data=0.028545044771738304), Valor(data=0.02873206292545772)]


In [148]:
resultado_final = []

for i in y_pred_teste:
    if i > Valor(0.5):
        resultado_final.append(1)
    else:
        resultado_final.append(0)

In [160]:
y_teste_lista = y_teste.tolist()


In [165]:
print(resultado_final)
print()
print(y_teste_lista)

[1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0]

[1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0]


In [198]:
precisao = precision_score(y_teste_lista, resultado_final)
acuracia = accuracy_score(y_teste_lista, resultado_final)
sensibilidade = recall_score(y_teste_lista, resultado_final)


print(f"Precisão: {precisao:.2f}")
print(f"Acurácia: {acuracia:.2f}")
print(f"Sensibilidade: {sensibilidade:.2f}")

Precisão: 1.00
Acurácia: 0.92
Sensibilidade: 0.83


A _precisão_ identifica se os dados classificados positivos são realmente positivos, importantes para identificar falsos positivos. Nesse caso, conseguimos identificar todos os tumores malignos!

A _acurácia_ nos retorna a proporção de acertos totais. Mantivemos uma medida de acerto acima de 90%, ótimo!

A _sensibilidade_ identifica entre os positivos reais, quantos foram identificados corretamente, importantes para identificar os falsos negativos.
Nessa caso, identificamos um tumor benigno como maligno, por isso nossa sensibilidade foi 83%

### REFERÊNCIAS:
1. Materiais disponibilizados pelo professor.
2. https://medium.com/ensina-ai/uma-explica%C3%A7%C3%A3o-visual-para-fun%C3%A7%C3%A3o-de-custo-binary-cross-entropy-ou-log-loss-eaee662c396c